scrapy_zyte_api/responses.py

```diff
@@ -0,0 +1,123 @@
+from base64 import b64decode
+from typing import Dict, List, Optional, Union, Tuple
+
+from scrapy import Request
+from scrapy.http import Response, TextResponse
+from scrapy.responsetypes import responsetypes
+
+_DEFAULT_ENCODING = "utf-8"
+
+
+class ZyteAPIMixin:
+
+    REMOVE_HEADERS = {
+        # Zyte API already decompresses the HTTP Response Body. Scrapy's
+        # HttpCompressionMiddleware will error out when it attempts to
+        # decompress an already decompressed body based on this header.
+        "content-encoding"
+    }
+
+    def __init__(self, *args, zyte_api: Dict = None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._zyte_api = zyte_api
+
+    @property
+    def zyte_api(self) -> Optional[Dict]:
+        """Contains the raw API response from Zyte API.
+
+        To see the full list of parameters and their description, kindly refer to the
+        `Zyte API Specification <https://docs.zyte.com/zyte-api/openapi.html#zyte-openapi-spec>`_.
+        """
+        return self._zyte_api
+
+    @classmethod
+    def _prepare_headers(cls, init_headers: Optional[List[Dict[str, str]]]):
+        if not init_headers:
+            return None
+        return {
+            h["name"]: h["value"]
+            for h in init_headers
+            if h["name"].lower() not in cls.REMOVE_HEADERS
+        }
+
+
+class ZyteAPITextResponse(ZyteAPIMixin, TextResponse):
+
+    attributes: Tuple[str, ...] = TextResponse.attributes + ("zyte_api",)
+
+    @classmethod
+    def from_api_response(cls, api_response: Dict, *, request: Request = None):
+        """Alternative constructor to instantiate the response from the raw
+        Zyte API response.
+        """
+        body = None
+        encoding = None
+
+        if api_response.get("browserHtml"):
+            encoding = _DEFAULT_ENCODING  # Zyte API has "utf-8" by default
+            body = api_response["browserHtml"].encode(encoding)
+        elif api_response.get("httpResponseBody"):
+            body = b64decode(api_response["httpResponseBody"])
+
+        return cls(
+            url=api_response["url"],
+            status=200,
```

Comment:
Shouldn't this be the site response status?
What if the site returns a 404 with valid content and we want to process that in the spider?

Reply:
It should, but Zyte API doesn't have a feature to return the status. It'll be implemented soon, but it's not ready yet.